### DISCUSSION TITLE
Websocket reconnection issues

### DISCUSSION THREAD

**Discussion by jeffective** [zauberzeug/nicegui#3726] [Q&A]
### Question

I am placing extreme load on my niecgui instance,  (many, many elements updating frequently).

This is resulting in random refreshing behavior in the clients, which I suspect is due to a websocket timeout issue.

How do I debug this issue or mitigate it?

I set the uvicorn log level to "info" and I can see the websocket connections being regularly closed / opened.

```
 ('', 57819) - "WebSocket /socket.io/?client_id=5cc7a195-02b2-46ba-bb1d-35d7b4ea8b8d&EIO=4&transport=websocket" [accepted]
INFO:     connection open
INFO:     connection closed
INFO:     ('', 55384) - "WebSocket /socket.io/?client_id=20a7e7c9-83a2-49f5-a423-48b7f6c8016f&EIO=4&transport=websocket" [accepted]
INFO:     connection open
INFO:     connection closed
INFO:     ('', 57822) - "WebSocket /socket.io/?client_id=5cc7a195-02b2-46ba-bb1d-35d7b4ea8b8d&EIO=4&transport=websocket" [accepted]
INFO:     connection open
INFO:     connection closed
```

**Comment by jeffective** [comment]
Issue might be localized to edge, difficult to reproduce on chrome

**Comment by rodja** [comment]
> websocket connections being regularly closed / opened

That is only a symptom. I wonder if it really is directly caused by "many, many elements updating frequently". It could also caused by blocking the main loop. Have you read [our FAQ about async/await](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates)? What do you see when activating asyncio debugging?

```py
def startup():
    loop = asyncio.get_running_loop()
    loop.set_debug(True)
    loop.slow_callback_duration = 0.05

app.on_startup(startup)
```

**Reply by jeffective** [reply]
thanks for the tip!!

I identified some tasks that are taking 50-65 ms to complete. But nothing more than that. However, these 50 ms tasks are part of a ui.update that was running at 4 Hz. So I'm thinking that these longer tasks are taking up too much cpu, which causes the small individual element updates to lag behind, triggering a refresh from the client?

**Reply by jeffective** [reply]
I also have multiple clients, about 3, so perhaps 50ms * 4 Hz * 3 clients = 600 ms / second of cpu is dangerously close to websocket timeout territory?

**Reply by jeffective** [reply]
increasing the reconnect timeout only seems to extend the time the gui freezes before refreshing

**Reply by jeffective** [reply]
I can induce the refresh by clicking on an interactive image many time, queuing up many actions that are being processed by the server, the server falls significantly behind, once all my clicks have gone through (over the course of a few seconds), the page freezes for approximately the reconnect timeout and refreshes

**Reply by jeffective** [reply]
I have my GUI pinned to one CPU and it seems to be using about 70-80% continuously for 3 clients.

**Reply by jeffective** [reply]
I can also trigger the refresh by clicking different ui.tabs a lot, each tab containing some large ui.tables and ui.interactive images

**Reply by rodja** [reply]
I suggest you further reduce the task slice execution times. I think there is no other way if you have such a load.

**Comment by jeffective** [comment]
I was able to reproduce the issue on the official nicegui documentation page for ui.interactive image
https://nicegui.io/documentation/interactive_image

by clicking about 7.5 times per second until the server fell far behind me and the page refreshed

**Reply by jeffective** [reply]
maybe the pong from the client is getting delayed by client side load?

**Reply by jeffective** [reply]
you have to click about a 1000 times but you can reproduce it

**Reply by jeffective** [reply]
ok maybe the client is sending pings that the server is not ponging fast enough?

**Reply by jeffective** [reply]
maybe related to client side websocket buffering

https://stackoverflow.com/questions/43998283/does-chrome-have-a-timeout-for-connected-web-sockets-that-have-not-sent-a-messag

**Reply by rodja** [reply]
We have not been able to reproduce the page refresh. 7.5 clicks per second is quite a hard to reach. But surely there will be some kind of limit what our server can handle at a given time.
